Feature: Plugin framework for Social Integration with Steam, Discord, GOG, etc

TrueBrain opened the pull request:
## Motivation / Problem

This is another approach to both #11391 and #8952 .
Closes #11391 
Closes #8952

Quality games in 2023 have integration with at least Steam, to show in more detail what a player is up to, and make it trivial to join their game (assuming it is open to join, ofc). OpenTTD is lacking this functionality.

The main hurdle has always been licensing; Steamworks SDK (and Discord SDK, and GOG SDK) are closed-source, and don't play nice with our GPLv2 license. So a simple integration was long ago established to be impossible, as that would give us a license-hell nobody wanted to deal with.

So in #8952 it was already suggested to create a plugin-API, where the API definition is licensed under MIT. This means that there can be other projects, with different licenses, that implement integration with platforms like Steam, Discord, GOG, ...
By doing it this way, it is very similar to content on BaNaNaS. Most of it is not licensed under GPLv2, yet any user can download it, and amend their game with all the goodies BaNaNaS content brings.

#8952 stalled in development, and a few months ago we got a similar PR #11391, trying to do the same thing. When looking at it this morning I was like: this can be done in a few days, and is actually not that much code. It is much more about figuring out how to deliver it to the clients, and how to make the experience as smooth as possible. So .. here we are.

## Description

Update:
Plugins are now validated whether they are signed by us. This, as otherwise we are afraid people will make misuse of this interface, and start running coin-miners or what-not on machine of users, by distributing a malicious but seemly useful plugin via the forums or whatnot.

Most of this work is heavily based and/or inspired by #8952, but with a few twists. First off, it assumes the plugins are distributed via BaNaNaS (more on this later). Second, it moves the code out of the "network" and into the global space, as it is not network specific. Lastly, by rewriting it myself I could clean up some bits to make it work with the code in 2023 (that PR was created in 2021).

I liked the approach of #8952:
- A plugin exports a single "init" function, which fills a table with pointers for OpenTTD to use. This is in contrast to #11391, which went for the route to export all functions OpenTTD might call. The benefit of the first approach is that a single plugin can implement several API versions, and staying compatible is a problem for the plugin; not for OpenTTD. In the second approach, we would have to add compatibility layers, to ensure older plugins still work.
- #11391 is event-based, and per event there is a payload. This makes for a bit of an awkward definition, and a lot can go wrong here. #8952 is much more verbose, and announces each function with their parameters (where the functions are just events). I liked this descriptive API definition much more.

This is a preference, and neither is actually better or worse than the other. There are a few other things that both did a bit weird in my opinion, so I took the opportunity to straiten that out too.

That all said, what is this actually doing:

On startup, the `social_integration` folder is scanned much like any other folder in OpenTTD. It follows the same rules, same searchpaths, etc. For every plugin found (they always end on `-social.(so|dll|dylib)`), it is loaded and the function `SocialPlugin_v1_GetInfo` and `SocialPlugin_v1_Init` is looked up. 

`GetInfo` is called first, which returns things like the name of the plugin, the social platform it is for, etc.
After, `Init` is called, which initializes the plugin. One of three things can happen:
- The plugin fails to initialize
- The plugin fails to initialize because, for example, Steam is not running
- The plugin is initialized

Now every tick the `RunCallbacks` is called. When it returns `false`, the plugin is unloaded. When OpenTTD quits, all plugins are unloaded.

Events that happen, like someone starting a single player, or joining a server, triggers an event (a function callback), to inform the plugin. The `RunCallbacks` is guaranteed to be called in the GameThread. All other callbacks (both directions) are undefined where they run, and no assumptions should be made. It is up to the plugin to be thread-safe for the functions called by OpenTTD (and up to OpenTTD to be thread-safe for the callbacks it gives the plugin).

The API itself is under `3rdparty/openttd_social_plugin_api`. This is done as `3rdparty` is where code lives under another license than GPL-v2. And the API definition is licensed under MIT, allowing plugins to use any license they like.

Social Presence plugins can be distributed in two ways:
- For Steam, GOG, etc, we prepackage them in the distribution.
- Via BaNaNaS (for later PRs)

This last one needs a bit more words:

Only developers are allowed to upload new content in this category on BaNaNaS. This is because these are executables we distribute, and so they have to be hold to the same standards as the game itself. Source in a Version Control System, build via a predefined CI workflow, and no human interaction during distribution. This to make the chance of someone sneaking in something bad very very small.
If anyone would be able to upload a plugin, it would be trivial for anyone to upload a coinminer, or worse. Other BaNaNaS content is protected from that, as we load those files ourselves and validate they are sane. But as the plugins are executables, this is impractical.

The upside of this approach is that we can update the plugins outside a normal OpenTTD release cycle; this means that if an SDK breaks something, we can quickly act, and don't have to wait for the next release. Also it means that older OpenTTD versions get updates for Steam, Discord, GOG, ... integration.

The downside is that I had to figure out how to pack both the plugin library as their dependency (all SDKs give their own library) in a way that OpenTTD could easily load. I succeeded on both Linux, MacOS, and Windows to do so.

Example of Steam and Discord plugins:
- https://github.com/OpenTTD/discord-social
- https://github.com/OpenTTD/steam-social
- https://github.com/OpenTTD/gog-galaxy-social

This PR deliberately is only adding a very small plugin interface. Later PRs can extend on it. But getting everything setup and working is already a huge task; so to not let it linger for a longer time because "oeh, we can also do NNN", the scope is made as small as possible, while still being useful.

For example, BaNaNaS support is prepared, but not part of this PR. This is because we need to figure things out like: there should only be one plugin for a social platform active; but what is "latest"?
For now, people can download plugins manually, and we will ship it with Steam / GOG automatically. Still a benefit to the user, while not getting overwhelmed with all the details we need to figure out for the next step.

![Screenshot 2023-12-29 202541](https://github.com/OpenTTD/OpenTTD/assets/1663690/79765199-caab-41bf-868d-8224ee798db6)
![Screenshot 2023-12-29 203120](https://github.com/OpenTTD/OpenTTD/assets/1663690/f4dc0607-97a0-4d65-bbb5-d74bd23bd31f)
![Screenshot 2023-12-29 205104](https://github.com/OpenTTD/OpenTTD/assets/1663690/ed5fddfa-937b-4aaf-b53a-0725555c0945)


## Limitations

- We, as developers, need to agree on this approach. As mentioned above, it is exactly the same as other BaNaNaS content, so this should be fine by anyones standard, but it is important to note that those plugins will not be under GPLv2 license.
- This is not supported on Emscripten, as it cannot communicate with Steam, Discord, GOG, ... anyway.
- Linux Distributions that want to distribute the plugins by compiling them theirselves, would need to sign them and add their public key to the list of keys in the game, before they can be loaded. As such, I suspect no distribution to do this.

## TODO list

- [x] Add to crashlog what libraries are loaded
- [x] Add to GUI somewhere what libraries are loaded
- [x] Create Discord Plugin
- [x] Create Steam Plugin
- [x] Create GOG Plugin
- [x] Track what the player is doing
  - [x] In main menu
  - [x] In scenario editor
  - [x] In singleplayer
  - [x] In multiplayer
- [x] Support Windows
- [x] Support Linux
- [x] Support MacOS

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

nielsmh commented on src/os/windows/library_loader_win.cpp:
```diff
@@ -0,0 +1,61 @@
+	auto error = GetLastError();
+	std::string message;
+	char buffer[512];
+	if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, error,
```
I worry that this might end up returning garbage for some locales. The error message might not be in English and the current ANSI codepage for the thread might not be UTF-8 (that might not even be supported on the Windows version the user is running), so this could end up generating mojibake when a non-UTF8 non-English string is treated as UTF-8 further down.

TrueBrain replied:
This is already existing code in our codebase; I just copied it as the other versions are baked into other flows (like network error handling). So it already stood the test of time.

Further more, the result is only used in Debug statements to allow figuring out why a library failed loading. So I am not too fuzzed about it.

nielsmh commented on src/lang/english.txt:
```diff
@@ -1076,6 +1078,19 @@ STR_GAME_OPTIONS_BASE_MUSIC                                     :{BLACK}Base mus
+STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_PLATFORM_NOT_RUNNING       :{ORANGE}{RAW_STRING} not running
+STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_UNLOADED                   :{RED}Unloaded
+STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_DUPLICATE                  :{RED}Duplicated plugin
+STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_UNSUPPORTED_API            :{RED}API not supported
```
Maybe it will be clearer as "Version not supported" or "Unsupported API version"?

nielsmh commented on src/library_loader.h:
```diff
@@ -0,0 +1,115 @@
+	 *
+	 * @param symbol_name The name of the symbol to get.
+	 */
+	void *GetSymbol(const std::string &symbol_name);
```
Would this make more sense as a string_view for the parameter? Since it's generally static string constants and not constructed values used in calls.

TrueBrain replied:
It cannot be a `std::string_view`, as in the usage, the strings need to be nul-terminated. Otherwise all implementations need to copy the string into a `std::string` to make it nul-terminated. Which ofc is a bit defeating the purpose.

This sadly happens when you interface with C-functions :(

So I think this is the best approach for us; if you have better suggestions, do let me know :)

nielsmh replied:
Aren't string literals in the form of `"foo"` specified to generate a char array like `{ 'f', 'o', 'o', '\0' }` i.e. the terminator is guaranteed to be part of the literal? But yes it would not give the guarantee if the strings were for some reason actual views of larger strings.

nielsmh commented on src/social_integration.h:
```diff
@@ -0,0 +1,84 @@
+	/**
+	 * Get the list of loaded social integration plugins.
+	 */
+	static std::vector<SocialIntegrationPlugin *> GetPlugins();
```
I think this can return a `const std::vector<SocialIntegrationPlugin *> &` instead, to avoid copying the vector every time someone wants the list just to inspect it.

rubidium42 commented on src/3rdparty/openttd_social_integration_api/openttd_social_integration_api_v1.h:
```diff
@@ -0,0 +1,154 @@
+	 */
+	const char *social_platform;
+
+	const char *name; ///< Full name of the plugin. UTF-8, nul-terminated. The plugin is and remains the owner of the memory.
```
Has any consideration been made to support translated strings?

Here `social_platform` seems to be most like the base set's name, and this `name` feels more like a base set's description. The base set's descriptions can be translated.

I'm aware that it is (currently) only for the name, as `social_platform` and `version` definitely shouldn't be translated I'd say. So, it might not be worth it. I just want to make sure a consideration has been made about it and it's not forgotten.

TrueBrain replied:
Yes, I did consider it! :)

Name/version of the Plugin are just that: names. Not a description, like with basesets. It is more to allow for competitive implementations of a single plugin, so we can distinguish them. And names like that don't need translation if you ask me.

Similar to social_platform, Steam is called Steam in any language. Equally, I suspect that a plugin name will be named the same in every language. For example, the current names are `Steam Integration`  and `Discord Integration` :) I have such an imagination :P

I did consider just removing `name` all together, as it doesn't actually add value. But again, it might be useful if we have more than one competing implementation of the same social platform.

rubidium42 commented on src/3rdparty/openttd_social_integration_api/openttd_social_integration_api_v1.h:
```diff
@@ -0,0 +1,154 @@
+	 * - GOG Galaxy
+	 * - ...
+	 */
+	const char *social_platform;
```
Maybe we should set up a wiki page that is the definitive source of social platform names, so when two people are working on XYZ they will both call it XYZ instead of one calling in Xyz. As that would defeat the whole preventing of duplicates. And by not putting it in the source code directly, later additions can still be easily managed.
Also maybe check for duplicate social platform names case insensitively?

TrueBrain replied:
Solved the last part. First part, sure, we can do that. Leaving open as I have yet to do that :)

Now done:
https://wiki.openttd.org/en/Development/Social%20Integration

rubidium42 commented on src/3rdparty/openttd_social_integration_api/openttd_social_integration_api.h:
```diff
@@ -0,0 +1,37 @@
+
+/* Although all the source-files created by OpenTTD are licensed under the
+ * GPL-v2, this file is an exception. This file is part of the API for
+ * social integration plugins, and licensed under MIT, to allow for non-free
```
```suggestion
 * social integration plugins, and licensed under the MIT license, to allow for non-free
```
and the few other places where this pre-amble is added.

rubidium42 commented on src/os/windows/library_loader_win.cpp:
```diff
@@ -0,0 +1,61 @@
+	std::string message;
+	char buffer[512];
+	if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, error,
+		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), buffer, sizeof(buffer), NULL) == 0) {
```
```suggestion
	if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, nullptr, error,
		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), buffer, sizeof(buffer), nullptr) == 0) {
```
Also, shouldn't we pass the lang ID from the language file here (`_current_language->winlangid`), so we get a translated error message back from Windows?

TrueBrain replied:
We don't in the other cases we use this function in a nearly identical way (which also use `NULL` btw :P Not for this PR to fix).

I think we talked about it during the development of the network variant of this function, and considered it to be better to use the system language, as then it matches the console language. Not sure, but I believe we spend time on this :)

Either way, if we want this, we should fix all instances of this function; so going to push that out of scope of this PR, if you don't mind :)

rubidium42 replied:
Since https://github.com/OpenTTD/OpenTTD/pull/11627 that changed; we now use `nullptr` in those cases.

rubidium42 commented on src/3rdparty/openttd_social_integration_api/LICENSE:
```diff
@@ -0,0 +1,20 @@
+ Copyright 2023 OpenTTD project
```
Should this year (and in the headers of the API) be reconsidered now it's 2024?

rubidium42 commented on src/openttd.cpp:
```diff
@@ -1009,6 +1012,28 @@ bool SafeLoad(const std::string &filename, SaveLoadOperation fop, DetailedFileTy
+			break;
+
+		case GM_NORMAL:
+			if (_network_server) {
```
```suggestion
			if (_networking) {
```
Or is playing as a client considered single player?

rubidium42 commented on src/os/unix/library_loader_unix.cpp:
```diff
@@ -0,0 +1,68 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file library_loader_win.cpp Implementation of the LibraryLoader for Linux / MacOS */
```
```suggestion
/** @file library_loader_unix.cpp Implementation of the LibraryLoader for Linux / MacOS */
```

rubidium42 commented on src/library_loader.h:
```diff
@@ -0,0 +1,115 @@
+/*
```
I think the library loader should at least be its own commit, if not its own PR.

TrueBrain replied:
Sounds like a good idea! This PR now includes [#11751](https://github.com/OpenTTD/OpenTTD/pull/11751).

rubidium42 reviewed (commented):
Also a tiny potential suggestion, depending on whether bananas integration is going to happen any time soon: a short manual on a wiki page listing known plugins and how to install them. That is linked from the social page when there are no plugins available.

PeterN commented on src/settings_gui.cpp:
```diff
@@ -166,6 +167,178 @@ static const std::map<int, StringID> _volume_labels = {
+		}
+
+		return *longest;
+	}
```
These need to use the longest pixel length, not the longest character length.

WWWWWWWW is wider than ..........

TrueBrain replied:
It would really help me if you tell how, instead of just: this is wrong :P

PeterN replied:
Normally I'd use `GetStringBoundingBox().width` in the window's `UpdateWidgetSize()` and return the width there, but pattern doesn't really work with your everything-in-a-custom-widget layout (which I really like btw).

So something like this, although maybe there's a std algorithm or some other pattern that can be used to reduce the duplication:
```
std::string *longest = &this->plugins[0]->version;
int longest_length = 0;
for (auto *plugin : this->plugins) {
    int length = GetStringBoundBox(plugin->version).width;
    if (length > longest_length) {
        longest_length = length;
        longest = &plugin->version;
    }
}

return *longest;
```

TrueBrain replied:
Awesome, tnx! Templating power will take care of the rest :)

rubidium42 commented on src/os/unix/unix.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "../../string_func.h"
 #include "../../fios.h"
 #include "../../thread.h"
+#include "../../debug.h"
```
Stray include.

2TallTyler commented on src/settings_gui.cpp:
```diff
@@ -166,6 +167,171 @@ static const std::map<int, StringID> _volume_labels = {
+						SetDParam(0, STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_UNSUPPORTED_API);
+						break;
+				}
+			} break;
```
`break` goes on its own line, no?

rubidium42 commented on src/settings_gui.cpp:
```diff
@@ -166,6 +167,171 @@ static const std::map<int, StringID> _volume_labels = {
+			case WID_GO_SOCIAL_PLUGIN_STATE: {
+				/* For SetupSmallestSize, use the longest string we have. */
+				if (this->current_index < 0) {
+					SetDParam(0, STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_UNSUPPORTED_API);
```
This is not necessarily the longest string in all translations. Actually "Discord not running" might even be a bit longer that "Unsupported version".
Practically you'd have to run through all the possibilities, i.e. all the states and all the plugins to determine the actual longest string. I'm not quite sure whether that's worth it though.

If the strings and the state are both in the same order, I'd just use the following to replace the big switch below:
```
SetDParam(0, STR_GAME_OPTIONS_SOCIAL_PLUGIN_STATE_RUNNING + plugin->state);
SetDParamStr(1, plugin->social_platform);
```

Then for getting the longest string, you can just iterate through the strings (by state) and set the platform, and simply try all cases. Yes, you will be setting the string unnecessarily, but I wonder if that's a big problem since it's not called that often.

TrueBrain replied:
I did a thing. Mostly because I really didn't feel like making this `State` enum iteratable with `_BEGIN` and `_END` stuff. It makes it so ugly. And also because I don't like `STR_... + i` stuff :P It always makes my code fail when I change things :D

So I went for a different approach ... but not sure what people think. The compiler likes it equally, as in, with `-O3` there is no difference between this and a switch-case. So that is something.

PeterN replied:
IMHO if you don't know what the longest string might be, and you don't want to test all possible combinations, it's better to design it around a wrapping mutli-line string instead. Especially if some translations might be quite long.

I guess the video driver description in the game options window is one example of where this is done.

TrueBrain replied:
For context, the string in question is `Steam is not running`. And I mostly got myself in this position as it is not a regular window, but a composed one :P
Wrapping would be a bit odd here, even for translations, in how it would look. So currently it makes the window wider. But I am not really sure this is the right approach.

Maybe I should just not try to put things this neatly, and just put it below each other, give it a width, and let wrapping happen when it needs to happen.

Meh :P

rubidium42 commented on src/settings_gui.cpp:
```diff
@@ -166,6 +167,171 @@ static const std::map<int, StringID> _volume_labels = {
+	{
+		this->plugins = SocialIntegration::GetPlugins();
+
+		if (this->plugins.size() == 0) {
```
```suggestion
		if (this->plugins.empty()) {
```

rubidium42 commented on src/social_integration.cpp:
```diff
@@ -0,0 +1,239 @@
+		Debug(misc, 1, "[Social Integration: {}] Loading ...", basepath);
+
+		auto &plugin = _plugins.emplace_back(std::make_unique<InternalSocialIntegrationPlugin>(filename, basepath));
+		UpdateExternalPlugins();
```
Can't this be called in `Scan()` after the `FileScanner::Scan`? Then we do not rebuild the list for each plugin. Now with 3 plugins, you need 6 inserts into the vector.

TrueBrain replied:
I would classify this as over-optimizing :) Pretty sure the list will never get long enough to show up on any performance benchmark :) I would rather keep it local to where it is important.

That said, I will see if I can rework it to be a bit nicer. But not for the performance :P More because it is a bit awkward how I did it :)

Edit: I made it pretty!

rubidium42 commented on src/social_integration.cpp:
```diff
@@ -0,0 +1,239 @@
+
+std::vector<SocialIntegrationPlugin *> &SocialIntegration::GetPlugins()
+{
+	return _external_plugins;
```
Personally I would have just created a local vector here, let that be filled from `_plugins` and then let return value optimisation take care to not introduce an actual copy of the vector. 

The current implementation will also work just fine.

TrueBrain replied:
You are not wrong; overoptimization on my side!

rubidium42 commented on src/social_integration.cpp:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	if (func != nullptr) {
+		func(args...);
```
Maybe slap some try-catch block around the call, so if it's written in C++ and throws an exception and doesn't catch it by itself, then that plugin doesn't take down the whole of OpenTTD with it.
```suggestion
	try {
		func(args...);
	} catch (...) {
		Debug(misc, 0, "Social plugin {} threw an exception", plugin->plugin_info.name);
	}
```

TrueBrain replied:
The API is exported as a pure C API, so we shouldn't be able to capture anything, I would think.

rubidium42 replied:
I took the steam-social plugin and replaced all functions content with `throw 1`. The result, this PR crashes hard:
`terminate called after throwing an instance of 'int'`

Now with a try-catch block around the functions (I counted at least 3) you call, I get the debug message instead. So, apparently if the C code in between is thin enough it works fine for at least my current compiler. That doesn't mean it'll always work fine, but then we tried our best to keep OpenTTD running and there's nothing more to do than to tell them to remove the plugin.

TrueBrain replied:
But in the same argument, if the plugin was in C, and crashes, it would still take the game down.

If (and that is a question: if) we want to solve this, I rather set a trap like we do in crashlog code, and capture any issue that might arise in the plugin handling.

But I think the question is here: do we actually want that? The plugins will be (after suggestion by 2TallTyler) guaranteed made available by us; so isn't it the same as a crash in the game, that we need to fix? Honest question, to be clear.
I just know from experience it is nearly impossible to prevent all ways a library can crash your application; so should we even try? Especially as a "throw" is rather unlikely to happen in the plugins.

TrueBrain replied:
I did some further reading, and my initial instinct is correct: C++ doesn't define the ABI of anything, let alone exceptions. This is why we fall back here ofc to a C ABI, as that is well defined.

It means what you tried basically only works between a single compiler, and MIGHT work between different compilers, but also might hopefully and utterly fail in worse ways than not doing it. Even different versions of the same compiler might handle the ABI differently, making a throw from the library incompatible with the application. As it is very likely the plugins will be compiled (over time) by other compilers and versions, it is not advised to send exceptions over the barrier.

A nice write-up I found about it here:

https://wiki.sei.cmu.edu/confluence/display/cplusplus/ERR59-CPP.+Do+not+throw+an+exception+across+execution+boundaries

Here they give a nice example that OpenTTD might be build in MinGW and the plugin with MSVC. This would absolutely not work, and MSVC uses SEH for exceptions, and MinGW does not. On Linux we might be a bit more lucky, but it is just that: luck. It is UB.

That was a nice trip down a rabbit hole which I actually enjoyed :)

What is left, I think, are two options:

- Use, as with crashlog code, a `setjmp` / SEH solution to trap any issues, and recover from it.
- Or, as because of the private/public key validation I will add, consider plugins as part of our codebase that are just loaded in at a later time, and that they are no different than any other routine in our codebase. They just live in another repo.

I still vote for the last, but if there is a good argument for the first, that can be done :)

2TallTyler commented on README.md:
```diff
@@ -118,6 +118,15 @@ Most types of add-on content can be downloaded within OpenTTD via the 'Check Onl
 Add-on content can also be installed manually, but that's more complicated; the [OpenTTD wiki](https://wiki.openttd.org/) may offer help with that, or the [OpenTTD directory structure guide](./docs/directory_structure.md).
 
 
+### 1.5.1) Social Integration
```
If I'm a privacy-conscious user downloading the game from Steam (cherry-picking when I care about my privacy, naturally 😛), I get the plugins installed automatically. It would be nice to have a sentence here telling me that if I delete the file, the plugin is gone. Just to sooth worried minds.

TrueBrain replied:
I am not sure.

Expecting Steam integration to be shipped with a game on Steam is kinda a no-brainer. Nobody is going to expect it to not be there, as most of the games on Steam ship it. If you are privacy conscious .. you would not have used Steam to download the game :D

For the Discord plugin on Steam, there you might have a point. I could make it a Steam DLC, if that helps.

The problem with deleting is that Steam will just restore it for you. So there is little point in suggesting something that is not going to work :)

2TallTyler replied:
Hmm, DLC could be a good way to distribute it. I agree that shipping Steam integration with a Steam build shouldn't bother anybody -- and it really does, they can download through openttd.org.

2TallTyler reviewed (commented):
Two thoughts here:

First, plugins (in their own repos) currently use hard-coded strings. I assume can we add them to eints?

Second, you talk about plugin safety for our official plugins distributed on BaNaNaS. But if I'm reading this correctly, it seems that OpenTTD will run any plugin dropped into the correct folder. I wonder if we should lock that down and allow only official plugins to run.

Let's say that we don't have an official plugin for the user's favorite social platform, say ~~Twitter~~ X. Or we someday discontinue support for a social platform because we disagree with a ToS update and the survey shows 0.1% of our players are using it anyway. Some enterprising person could create their own plugin that includes a crypto-miner or keylogger or whatever, advertise it through Discord or GitHub or TT-Forums, and get a bunch of users to download it. Those players are accustomed to OpenTTD mods being safe to download and use, because of the sandbox that GRFs and AI/GS scripts are contained within. Not the case here. We didn't distribute the malicious code, sure, but if that user runs it through OpenTTD... do we share any of that responsibility?

TrueBrain committed:
Add: signature validation before loading plugins

This prevents us from executing unauthorized plugins.

TrueBrain commented on src/signature.cpp:
```diff
@@ -0,0 +1,290 @@
+/** The public keys used for signature validation. */
+static std::vector<std::string> _public_keys_v1 = {
+	"1234567890123456789012345678901234567890123456789012345678901234",
+	"1234567890123456789012345678901234567890123456789012345678901234",
```
Ofc these are not valid yet; I need to generate a private/public pair and add the private part to the CI where needed. And the public pair here.

TrueBrain commented on src/social_integration.cpp:
```diff
@@ -25,16 +26,23 @@
 	{
 		openttd_info.openttd_version = _openttd_revision;
+
+		if (!ValidateSignatureFile(fmt::format("{}.sig", filename))) {
```
For development builds, it should be possible to bypass this check, and just make it continue.

TrueBrain committed:
Add: API definition of our upcoming social integration API

TrueBrain committed:
Add: scaffolding for finding social plugins on disk

TrueBrain committed:
Add: initialize plugin when found

TrueBrain committed:
Add: update README with instructions on social plugin

TrueBrain committed:
Add: don't allow two plugins for the same social platform

The second one will be unloaded as soon as it is detected it is
for the same platform.

TrueBrain committed:
Add: make some parts of the plugin information available for public use

TrueBrain committed:
Add: Social Plugin information in the survey / crashlog

TrueBrain committed:
Add: new tab Social in Game Options to show the loaded plugins

TrueBrain committed:
Add: different callbacks to update the Social Plugins at the right time

glx22 commented on src/3rdparty/openttd_social_integration_api/openttd_social_integration_api.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2023 OpenTTD project
```
```suggestion
 * Copyright 2024 OpenTTD project
```

glx22 commented on src/3rdparty/openttd_social_integration_api/openttd_social_integration_api_v1.h:
```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2023 OpenTTD project
```
```suggestion
 * Copyright 2024 OpenTTD project
```

glx22 reviewed (commented):
Just to match the standalone license file.

rubidium42 commented on src/social_integration.cpp:
```diff
@@ -40,6 +59,45 @@ class SocialIntegrationFileScanner : FileScanner {
+
+			case OTTD_SOCIAL_INTEGRATION_V1_INIT_PLATFORM_NOT_RUNNING:
+				Debug(misc, 0, "[Social Plugin: {}] Failed to initialize: {} is not running", basepath, plugin->plugin_info.social_platform);
+				return false;
```
The begin of the  `Debug` messages feel a bit inconsistent. Or is it wanted that messages before attempting the initialisation describe the plugin differently than after?
The first two are "Social Integration: <name>", the last three are "Social Plugin: <name>" but that might be intended.
The third is like the first two but missing a colon; this is definitely triggering my OCD.

TrueBrain replied:
I am happy with your OCD, as they were all meant to be the same :D

rubidium42 commented on src/signature.cpp:
```diff
@@ -0,0 +1,328 @@
+	}
+
+	return true;
+}
```
settings.cpp has `DecodeHexNibble` and `DecodeHexText`. Maybe it should be moved to string.cpp/string_func.h, so it they can share the same function.
Found out about their existence when looking at PR 11855, because it changed the NewGRF configuration output.

TrueBrain replied:
Yeah, I copied the Nibble from it. The DecodeHexText has a very very odd return value. I might refactor that first, as it just hurts my eyes :)

rubidium42 commented on src/signature.cpp:
```diff
@@ -0,0 +1,328 @@
+	 *   {
+	 *     "files": [
+	 *       {
+	 *         "checksum": "version$hash"
+	 *         "filename": "filename",
+	 *       },
+	 *       ...
+	 *     ],
+	 *     "signature": "version$signature"
```
What's the reasoning for putting version into the checksum/signature string, and not as a separate field? Now you need to manually split the string, instead of reading two fields.

TrueBrain replied:
Unix. I like their passwd which also includes the version. If you copy the string now, it is obvious what version it refers to, and there is no question. But it isn't a strong reason.

Will give it some thought if I want to keep it like this or not.

TrueBrain replied:
Gave it some more thought, and I think this is the right way to go (what is currently done).

If I would do it in a separate field, you would get something like `checksum_version`, which feels really awkward and bulky. And I don't want to do a `schema` or something, as I suspect that it is more likely that one of the two (signature or checksum) changes, than both at the same time. This is more descriptive long-term, I think.

The other solution would be to nest the checksum, so you get:

```
{
  "checksum": {
    "version": 1,
    "checksum": "...",
  }
}
```

But I dunno .. lot of letters for basically the same thing.

Always tricky, designing these things with long-term in mind :D

rubidium42 commented on src/signature.cpp:
```diff
@@ -0,0 +1,328 @@
+{
+	return _ValidateSignatureFile(filename);
+}
+#endif
```
I would put the `#if`, `#else` and `#endif` within the function, so there is only one function definition.

rubidium42 reviewed (commented):
Can't find anything major wrong, just some minutiae.
